qiskit/transpiler/exceptions.py

```diff
@@ -15,16 +15,25 @@
 """
 Exception for errors raised by the transpiler.
 """
+import warnings
 from qiskit.exceptions import QiskitError
 
 
-class TranspilerError(QiskitError):
-    """Exceptions raised during transpilation"""
-
-
 class TranspilerAccessError(QiskitError):
     """Exception of access error in the transpiler passes."""
 
+    def __init__(self, *message):
+        """Set the error message."""
+        super().__init__(' '.join(message))
+        warnings.warn('The exception TranspilerAccessError is deprecated as of 0.11.0 '
+                      'and will be removed no earlier than 3 months after that release '
+                      'date. You should use the exception TranspilerError instead.',
+                      DeprecationWarning, stacklevel=2)
```

Comment:
One question about this (sorry I didn't think about it before). Won't this raise an error when a `TranspilerError` object is initialized too? Since TranspilerError is a subclass for backwards compatibility during the deprecation period its `__init__()` will call the parent classes by default. This means we'll end up with a warning even if TranspilerError is used. We would have had the same problem with the decorator before. 

The only way I can think handling this is either to check if the class is TranspilerError in the `__init__` and if so don't raise a warning. But I'm not sure if that would actually be what we want. The deprecation message here is presumably to warn users that are still catching TranspilerAccessError outside of qiskit that in the future that will be changing. If we raise a warning every time TranspilerAccessError is created that seems extra noise because the end user may not actually need to know. Or am I just overthinking things. :)

What do you think?

Reply:
I also see this as a problem but I dont think that a solution is that simple.

In my opinion the desired behaviour is as follows: If the TranspilerError is triggered normally, no deprecation warning should be raised. In the case a TranspilerError is triggered but a TranspilerAccessError is expected, the deprecation warning should be raised. Additionally, if a TranspilerAccessError is raised the deprecation warning should be raised too. 

I will think about a solution for this problem. In my opinion this also affects other issues like #3213 or #3206.

Reply:
An additional problem is that if TranspilerAccessError is not instantiated like in an isinstance check, the deprecation warning is not issued. But in my opinion this is one of the typical cases we want this warning to be issued. 

The best way would be to issue a warning in case someone imports the TranspilerAccessError explicitly.